google/cloud/internal/async_long_running.h

```diff
@@ -0,0 +1,173 @@
+      location);
+  if (!status.ok()) return status;
+  return result;
+}
+
+/**
+ * Asynchronously starts and polls a long-running operation.
+ *
+ * Long-running operations [aip/151] are used for API methods that take a
+ * significant amount of time to complete (think minutes, maybe an hour). The
+ * gRPC API returns a "promise" object, represented by the
+ * `google::longrunning::Operation` proto, and the application (or client
+ * library) should periodically poll this object until it is "done".
+ *
+ * In the C++ client libraries we represent these long-running operations by
+ * a member function that returns `future<StatusOr<ReturnType>>`. This function
+ * is a helper to implement these member functions. It first starts the
+ * operation using an asynchronous retry loop, and then starts an asynchronous
+ * loop to poll the operation until it completes.
+ *
+ * The promise can complete with an error, which is represented by a
+ * `google::cloud::Status` object, or with success and some `ReturnType` value.
+ * The application may also configure the "polling policy", which may stop the
+ * polling even though the operation has not completed.
+ *
+ * Library developers would use this function as follows:
+ *
+ * @code
+ * class BarStub {
+ *  public:
+ *   virtual future<StatusOr<google::longrunning::Operation>> AsyncFoo(
+ *     google::cloud::CompletionQueue& cq,
+ *     std::unique_ptr<grpc::ClientContext> context,
+ *     FooRequest const& request) = 0;
+ *
+ *   virtual future<StatusOr<google::longrunning::Operation>> AsyncGetOperation(
+ *     google::cloud::CompletionQueue& cq,
+ *     std::unique_ptr<grpc::ClientContext> context,
+ *     google::longrunning::GetOperationRequest const& request) = 0;
+ * };
+ * @endcode
+ *
+ * The corresponding `*ConnectionImpl` class would look as follows:
+ *
+ * @code
+ * class BarConnectionImpl : public BarConnection {
+ *  public:
+ *   // Using C++14 for exposition purposes. The implementation supports C++11.
+ *   future<StatusOr<FooResponse>> Foo(FooRequest const& request) override {
+ *     return google::cloud::internal::AsyncLongRunningLoop(
+ *       cq_, request,
+ *       [stub = stub_](auto& cq, auto context, auto const& request) {
+ *         return stub->AsyncFoo(cq, std::move(context), request);
+ *       },
+ *       [stub = stub_](auto& cq, auto context, auto const& request) {
+ *         return stub->AsyncGetOperation(cq, std::move(context), request);
+ *       },
+ *       retry_policy_->clone(), backoff_policy_->clone(),
+ *       IdempotencyPolicy::kIdempotent,
+ *       polling_policy_->clone(),
+ *       __func__ // for debugging
+ *       );
+ *   }
+ *
+ *  private:
+ *    google::cloud::CompletionQueue cq_;
+ *    std::shared_ptr<BarStub> stub_;
+ * };
+ * @endcode
+ *
+ * [aip/151]: https://google.aip.dev/151
+ */
+template <typename ReturnType, typename RequestType, typename StartFunctor,
+          typename RetryPolicyType>
```

Comment:
Why is `RetryPolicyType` a template parameter, but `BackoffPolicy` is just a concrete type and not templated? Maybe I just forget why this is like this?

Reply:
There is no common base class for the Bigtable retry policies:

https://github.com/googleapis/google-cloud-cpp/blob/0687743d1881cb458c0b11a043a7d72401c07f56/google/cloud/bigtable/rpc_retry_policy.h#L81

and all the other retry policies:

https://github.com/googleapis/google-cloud-cpp/blob/0687743d1881cb458c0b11a043a7d72401c07f56/google/cloud/internal/retry_policy.h#L33